### PR TITLE
Adds a macro language for defining, parsing, resolving, and matching against paths. These will be used as part of the file system overhaul for Project definitions of pathing.

### PR DESCRIPTION
**Pull request by SavagePencil**
THIS IS INERT CODE. BIGGER CODE CHANGES FOR THE PROJECT MANAGER, PROJECT FILE SCHEMA, AND PROJECT SCENARIO TEMPLATES TO COME.

## Summary

Implements a comprehensive macro language parser for template-based path generation and matching in Griptape Nodes. This enables dynamic path resolution with variable substitution, format specifications, and optional variables.

### Core Functionality
- **Template Parsing**: Parse macro templates with `{variable}` syntax
- **Path Generation**: Resolve templates to concrete paths using variable values
- **Path Matching**: Extract variable values from paths by matching against templates (greedy matching strategy)

### Template Syntax
- Required variables: `{variable_name}`
- Optional variables: `{variable_name?}`
- Separator format: `{variable:_}` (e.g., `my_value`)
- Numeric padding: `{frame:03}` (e.g., `005`)
- Case transformations: `{name:lower}`, `{name:upper}`, `{name:slug}`
- Multiple format specs: `{name:_:lower}` (chained transformations)
- Environment variables: Resolved via SecretsManager integration

### API Design

**ParsedMacro class** provides:
- `resolve()` - Generate paths from templates and variables
- `matches()` - Check if a path matches the template (returns bool)
- `extract_variables()` - Extract variable values from paths (returns dict with string keys)
- `find_matches_detailed()` - Extract variables with metadata (returns dict with VariableInfo keys)
- `get_variables()` - Get list of all variables in template

### Implementation Highlights

1. **Greedy Matching Strategy**: Returns single results instead of exploring all possible interpretations
2. **Partial Resolution**: Resolves known variables before matching to reduce ambiguity
3. **Format Spec Reversal**: Can reverse transformations (e.g., `"005"` → `5` for numeric padding)
4. **Optional Variable Handling**: Optionals not in `known_variables` are automatically removed during partial resolution
5. **Clean OOP Design**: All functionality unified in `ParsedMacro` class (eliminated separate Parser/Resolver classes)

## Architecture

- **`ParsedMacro`**: Main API class for parsing, resolving, and matching
- **`_PartiallyResolvedMacro`**: Internal state for partial resolution (private)
- **Format Specs**: `SeparatorFormat`, `NumericPaddingFormat`, `LowercaseFormat`, `UppercaseFormat`, `SlugFormat`
- **Segment Types**: `ParsedVariable`, `ParsedStaticValue` for representing parsed templates

## Testing

- 58 comprehensive unit tests covering all functionality
- Tests for parsing, resolution, matching, format specs, and edge cases
- All tests passing

## Type Safety

- Full type annotations throughout
- Passes strict type checking (pyright)
- No type errors or warnings

### CODE REVIEW

File: src/griptape_nodes/common/macro_parser/formats.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+@dataclass
+class DateFormat(FormatSpec):
+    """Date formatting like :%Y-%m-%d."""
+
+    pattern: str  # e.g., "%Y-%m-%d"
+
+    def apply(self, _value: str | int) -> str:
+        """Apply date formatting."""
+        # TODO(https://github.com/griptape-ai/griptape-nodes/issues/XXXX): Implement date formatting
```
Comment (collindutter) [inline src/griptape_nodes/common/macro_parser/formats.py]: Pls make real TODOs

**Review by collindutter — CHANGES_REQUESTED** [review]

File: src/griptape_nodes/common/macro_parser/formats.py (hunk 1 above)
Comment (SavagePencil) [reply src/griptape_nodes/common/macro_parser/formats.py]: oh jeez! I'll just implement them.

File: src/griptape_nodes/common/macro_parser/formats.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        msg = "DateFormat not yet fully implemented"
+        raise MacroResolutionError(msg)
+
+    def reverse(self, value: str) -> str:
+        """Attempt to parse date string."""
+        # TODO(https://github.com/griptape-ai/griptape-nodes/issues/XXXX): Implement date parsing
```
[the remark this answers is not in the document]
Comment (SavagePencil) [reply src/griptape_nodes/common/macro_parser/formats.py]: lemme just do it
